### ISSUE TITLE
Excessive terms clearing during the partial purge

### ISSUE THREAD

**Issue by piotrbak** [wp-media/wp-rocket#2549]
**Describe the bug**
We're clearing taxonomy terms recursively during the partial purge.

**To Reproduce**
1. Assign a Woo product to the category A
2. Create categories B, C, D that are children of the category A
3. Make sure that the product is not present in the child categories
4. Update the product
5. We'll pass `domain.com/a/` to the `rocket_rrmdir` via this part of code:
https://github.com/wp-media/wp-rocket/blob/00cf7a1c5e0b2e7354b0aa76ca86603ee4420eae/inc/common/purge.php#L138
6. It will also clear  `domain.com/a/b/` `domain.com/a/c/` `domain.com/a/d/` and all further nested categories where the product is not present.

**Expected behavior**
Make sure that child terms are not cleared if the post is not added there.

**Additional context**
It could be a pretty big problem if products are sharing category permalinks like following:
`domain.com/product/` - Category base permalink
`domain.com/product/%product_cat%/` - Product permalink
The above will cause the purge of all products that are present in the passed category term. However, Remy mentioned that the above example is an edge case:
https://wp-media.slack.com/archives/C43T1AYMQ/p1585602235046800?thread_ts=1585337918.031600&cid=C43T1AYMQ

Related ticket:
https://secure.helpscout.net/conversation/1113270980/152492?folderId=2415573

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by Tabrisrp** [comment]
**Reproduce the issue** ✅ 
**Identify the root cause** ✅ 
Piotr handled those already 👍 

**Scope a solution** ❌
This one is tough, because the system is not currently designed to handle directories to preserve on that part.

We could only delete the files in the corresponding cache folder, but that would ignore any special folder like the one created when there is cached query strings, so that's not ok.

@hellofromtonya maybe we can discuss together how best to approach that

**Comment by hellofromtonya** [comment]
Good idea @Tabrisrp. Let's pull this into Slack and figure out the best way to handle this.

**Comment by Tabrisrp** [comment]
**Scope a solution** ✅ 
Encapsulate getting the term(s) URL(s) and the appropriate files to clean in a private function dedicated to that. 

The function needs to get any files inside the provided term cache directory, following the pattern `index(.*).html(.*)`, and also any directory starting with `?` indicating it's a cached query string directory.

Use this function in `rocket_clean_term()` and `rocket_clean_post()` to prevent the recursive issue

**Effort** ✅ 
Effort `[S]`

**Comment by engahmeds3ed** [comment]
Blocked by #2750

**Comment by GeekPress** [comment]
@engahmeds3ed @hellofromtonya As #2750 has been merged, I suppose this issue isn't blocked anymore?

**Comment by arunbasillal** [comment]
Related ticket - https://secure.helpscout.net/conversation/1287845419/196367/

**Comment by engahmeds3ed** [comment]
@arunbasillal 
I was trying to reproduce this issue locally but I think the code changed a lot so I think this is working normally now.

![image](https://user-images.githubusercontent.com/15707971/99262598-d9084c00-2826-11eb-9582-b5066c341abe.png)

and after updating the product `product-level1` which is related to the main parent category `level1` this is the result

![image](https://user-images.githubusercontent.com/15707971/99262716-f9d0a180-2826-11eb-9aab-c7e96ce21108.png)

and the cleaned products were because of we get next/previous on the same category and out of the same category.

So I think it's working now and when talked with @piotrbak he said that he needs to validate more to be sure that it's solved.

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1800682184/328667/

**Comment by DahmaniAdame** [comment]
Related - [https://secure.helpscout.net/conversation/1816327961/331699/](https://secure.helpscout.net/conversation/1816327961/331699/)

**Comment by thinkjarvisdesignandmarketing** [comment]
@piotrbak @arunbasillal 
Did this issue get resolved?

We have an extremely large ecommerce site and are considering using an alternative caching system because of this problem. Orders are placed every 5-15 minutes. I only really want the cache to clear on the post itself not the surrounding categories.

**Comment by piotrbak** [comment]
Hello @thinkjarvisdesignandmarketing unfortunately it's still not fixed.

`I only really want the cache to clear on the post itself not the surrounding categories.`
You could unhook our clearing functions and hook your own to clear whatever is needed. If you have troubles with implementation, please reach our Support Team.

**Comment by thinkjarvisdesignandmarketing** [comment]
> Hello [@thinkjarvisdesignandmarketing](https://github.com/thinkjarvisdesignandmarketing) unfortunately it's still not fixed.
> 
> `I only really want the cache to clear on the post itself not the surrounding categories.` You could unhook our clearing functions and hook your own to clear whatever is needed. If you have troubles with implementation, please reach our Support Team.

Thanks @piotrbak I am struggling with this to be honest.

We could do with a set of snippets in the WP-Rocket documentation for particular use cases.

I can see the helper plugin to disable the function completely.

Im also currently investigating using Ajax to grab the stock values, therefore the shop stays 100% cached without any stock sync issues.

**Comment by girlie** [comment]
Related: https://secure.helpscout.net/conversation/2779821840/527898/

**Comment by viobru** [comment]
Related: https://secure.helpscout.net/conversation/3117323204/594822
